# Data Wrangling - WeRateDogs Twitter Data

In [2]:
import pandas as pd 
import numpy as np
import json
import requests
import tweepy 
import os

## Gather Data
1. The WeRateDogs Twitter archive.
2. The tweet image predictions from Udcaity servers.
3. The retweet count and faorite count from tweepy API

In [3]:
# 1. Read the WeRateDogs Twitter archive as DataFrame
df_tweets = pd.read_csv("twitter-archive-enhanced.csv")

In [ ]:
# 2. Download the tweet image predictions from Udcaity servers.
url = "https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"
response = requests.get(url)

with open(os.path.join("image_predictions.tsv"), mode = 'wb') as file:
    file.write(response.content)

In [4]:
# Read the image predictions as DataFrame
df_image = pd.read_csv('image_predictions.tsv', sep = '\t')

In [ ]:
# 3. Get the retweet count and faorite count from tweepy API
consumer_key = ''
consumer_secret = ''
access_token = ''
access_secret = ''
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)
api = tweepy.API(auth, wait_on_rate_limit = True, wait_on_rate_limit_notify = True)


In [ ]:
# Write API data to tweet_json.txt file
count = 0
with open(os.path.join("tweet_json.txt"), mode = 'w', encoding="utf8") as file:
    for t_id in tweets["tweet_id"]:
        try:
            file.writelines(json.dumps(api.get_status(t_id, tweet_mode='extended')._json) + "\n")
            count = count +1
            print(str(count) + ": The id is " + str(t_id))
        except: # catch exceptions
            print(str(count) + ": There is no id for " + str(t_id))

In [5]:
# Load json data from tweet_json.txt
with open('tweet_json.txt', encoding="utf8") as f:
    all_json = [json.loads(line) for line in f]

In [6]:
# Get tweet_id, retweet_count, favorite_count and make a DataFrame
df_tweets_api = pd.DataFrame({
    'id' : list(map(lambda json: json['id'] , all_json)),
    'retweet_count' : list(map(lambda json: json['retweet_count'] , all_json)),
    'favorite_count' : list(map(lambda json: json['favorite_count'] , all_json))
})


## Assess Data
#### `df_tweets` DataFrame Analysis

In [24]:
# Assess the dw_tweets visually
df_tweets

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193 NaN                    NaN                    
1     892177421306343426 NaN                    NaN                    
2     891815181378084864 NaN                    NaN                    
3     891689557279858688 NaN                    NaN                    
4     891327558926688256 NaN                    NaN                    
5     891087950875897856 NaN                    NaN                    
6     890971913173991426 NaN                    NaN                    
7     890729181411237888 NaN                    NaN                    
8     890609185150312448 NaN                    NaN                    
9     890240255349198849 NaN                    NaN                    
10    890006608113172480 NaN                    NaN                    
11    889880896479866881 NaN                    NaN                    
12    889665388333682689 NaN                    NaN                    
13    889638837579907072 NaN                    NaN                    
14    889531135344209921 NaN                    NaN                    
15    889278841981685760 NaN                    NaN                    
16    888917238123831296 NaN                    NaN                    
17    888804989199671297 NaN                    NaN                    
18    888554962724278272 NaN                    NaN                    
19    888202515573088257 NaN                    NaN                    
20    888078434458587136 NaN                    NaN                    
21    887705289381826560 NaN                    NaN                    
22    887517139158093824 NaN                    NaN                    
23    887473957103951883 NaN                    NaN                    
24    887343217045368832 NaN                    NaN                    
25    887101392804085760 NaN                    NaN                    
26    886983233522544640 NaN                    NaN                    
27    886736880519319552 NaN                    NaN                    
28    886680336477933568 NaN                    NaN                    
29    886366144734445568 NaN                    NaN                    
...                  ...  ..                     ..                    
2326  666411507551481857 NaN                    NaN                    
2327  666407126856765440 NaN                    NaN                    
2328  666396247373291520 NaN                    NaN                    
2329  666373753744588802 NaN                    NaN                    
2330  666362758909284353 NaN                    NaN                    
2331  666353288456101888 NaN                    NaN                    
2332  666345417576210432 NaN                    NaN                    
2333  666337882303524864 NaN                    NaN                    
2334  666293911632134144 NaN                    NaN                    
2335  666287406224695296 NaN                    NaN                    
2336  666273097616637952 NaN                    NaN                    
2337  666268910803644416 NaN                    NaN                    
2338  666104133288665088 NaN                    NaN                    
2339  666102155909144576 NaN                    NaN                    
2340  666099513787052032 NaN                    NaN                    
2341  666094000022159362 NaN                    NaN                    
2342  666082916733198337 NaN                    NaN                    
2343  666073100786774016 NaN                    NaN                    
2344  666071193221509120 NaN                    NaN                    
2345  666063827256086533 NaN                    NaN                    
2346  666058600524156928 NaN                    NaN                    
2347  666057090499244032 NaN                    NaN                    
2348  666055525042405380 NaN                    NaN                    
2349  666051853826850816 NaN                    NaN     

Note:
1. There are retweet data that starts with **RT** in `text` column
2. Odd dog `name` with **a**
3. Empty `expanded_urls`
4. None for NaN value

In [25]:
# Assess the df_tweets programmatically
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [26]:
df_tweets.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03  7.800000e+01           7.800000e+01          
mean   7.427716e+17  7.455079e+17           2.014171e+16          
std    6.856705e+16  7.582492e+16           1.252797e+17          
min    6.660209e+17  6.658147e+17           1.185634e+07          
25%    6.783989e+17  6.757419e+17           3.086374e+08          
50%    7.196279e+17  7.038708e+17           4.196984e+09          
75%    7.993373e+17  8.257804e+17           4.196984e+09          
max    8.924206e+17  8.862664e+17           8.405479e+17          

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count  1.810000e+02         1.810000e+02              2356.000000        
mean   7.720400e+17         1.241698e+16              13.126486          
std    6.236928e+16         9.599254e+16              45.876648          
min    6.661041e+17         7.832140e+05              0.000000           
25%    7.186315e+17         4.196984e+09              10.000000          
50%    7.804657e+17         4.196984e+09              11.000000          
75%    8.203146e+17         4.196984e+09              12.000000          
max    8.874740e+17         7.874618e+17              1776.000000        

       rating_denominator  
count  2356.000000         
mean   10.455433           
std    6.745237            
min    0.000000            
25%    10.000000           
50%    10.000000           
75%    10.000000           
max    170.000000

Note:
 1. `rating_numerator` and `rating_denominator` have too large max value
 2. `timestampe` might need to change data type

In [ ]:
df_tweets['rating_denominator'].value_counts()

In [ ]:
df_tweets['rating_numerator'].value_counts()

In [27]:
# Check recordes with rating denominator not equals to 10
pd.set_option('display.max_colwidth', -1)
df_tweets[df_tweets['rating_denominator'] > 10][['tweet_id', 'text', 'rating_numerator', 'rating_denominator', 'name']]

tweet_id  \
342   832088576586297345   
433   820690176645140481   
784   775096608509886464   
902   758467244762497024   
1068  740373189193256964   
1120  731156023742988288   
1165  722974582966214656   
1202  716439118184652801   
1228  713900603437621249   
1254  710658690886586372   
1274  709198395643068416   
1351  704054845121142784   
1433  697463031882764288   
1598  686035780142297088   
1634  684225744407494656   
1635  684222868335505415   
1662  682962037429899265   
1663  682808988178739200   
1779  677716515794329600   
1843  675853064436391936   

                                                                                                                                                    text  \
342   @docmisterio account started on 11/15/15                                                                                                             
433   The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd                                                  
784   RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…         
902   Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE                                                                       
1068  After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ         
1120  Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv                            
1165  Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a                                                                                     
1202  This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq                                
1228  Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1                                                        
1254  Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12                                         
1274  From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK   
1351  Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa                                                                   
1433  Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ                                                         
1598  Yes I do realize a rating of 4/20 would've been fitting. However, it would be unjust to give these cooperative pups that low of a rating             
1634  Two sneaky puppers were not initially seen, moving the rating to 143/130. Please forgive us. Thank you https://t.co/kRK51Y5ac3                       
1635  Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110 https://t.co/1zfnTJLt55         
1662  This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5          
1663  I'm aware that I could've said 20/16, but here at WeRateDogs we are very professional. An inconsistent rating scale is simply irresponsible          
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think https://t.co/ZanVtAtvIq                                                                               
1843  Here we have an entire platoon of puppers. Total score: 88/80 would pet all at once https://t.co/y93p6FLvVw                                          

      rating_numerator  rating_denominator      name  
342   11                15                  None      
433   84                70              

In [ ]:
# Check recordes with rating numerator are not in normal range
df_tweets[(df_tweets['rating_numerator'] > 20) | (df_tweets['rating_numerator'] < 10)][['tweet_id', 'text', 'rating_numerator', 'rating_denominator', 'name']]

Note:
1. Some `rating_numerator` values are wrong when encountering decimal values. 
2. Sometimes there are multiple matches for Number/Number in the text.
3. Some numerators are not 10.

In [ ]:
df_tweets['source'].value_counts()

Note:
 - The rating_numerator and rating_denominator in `df_tweets` should have further look afterwards. 

In [ ]:
len(df_tweets[df_tweets.duplicated()])

`df_tweets` has no duplicate records

#### `df_image` DataFrame Analysis

In [ ]:
df_image.sample(20)

In [ ]:
df_image.info()

In [ ]:
len(df_image[df_image.duplicated()])

In [ ]:
df_image.describe()

In [ ]:
df_image[df_image['p1_dog'] == True]['p1'].value_counts()

In [ ]:
df_image[(df_image['p1_dog'] == False) & (df_image['p2_dog'] == True)]['p2'].value_counts()

In [ ]:
df_image[(df_image['p1_dog'] == False) & (df_image['p2_dog'] == False) & (df_image['p3_dog'] == True)]['p3'].value_counts()

#### `df_tweets_api` DataFrame Analysis

In [ ]:
df_tweets_api.sample(20)

In [ ]:
df_tweets_api.info()

In [ ]:
len(df_tweets_api[df_tweets_api.duplicated()])

#### Quality 
##### `df_tweets` DataFrame
- Retweet data shouldn't be included
- Incorrect `rating_numerator` and `rating_denominator`
- **Doggo, floofer, pupper, puppo** columns should have NaN, not None when it is null
- Inaccurate dog **name**: such, quite, an, a
- Dog **name** should have NaN rather than None when it is null
- Timestamp column should be datetime type
- Missing **expanded_urls** (2,297 instead of 2,356)

##### `df_image` DataFrame
- Missing records (2,075 instead of 2,356)
- **p1, p2, p3** should have consistent capitalization

##### `df_tweets_api` DataFrame
- Missing records (2,343 instead of 2,356)

#### Tidiness
- The prediction of `df_image` DataFrame should be part of the tweets table
- Favorite count and retweet count in `df_tweets_api` DataFrame should be part of the tweets table
- **retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp** should be removed since they are for retweet info
- **Doggo, floofer, pupper, puppo** columns should combine into one column **dog_stage**
- **Timestamp** could have extra columns **year, month, weekofDay** to analyze post trends in different time period

## Clean

In [7]:
df_tweets_copy = df_tweets.copy()
df_image_copy = df_image.copy()
df_tweets_api_copy = df_tweets_api.copy()

### Tidiness
1. `df_image` and `df_tweets_api` DataFrame should be part of the `df_tweets` DataFrame
2. Drop extra retweet columns
3. Drop extra dog stage columns

##### Define
1. Merge `df_tweets`, `df_image` and `df_tweets_api` on same tweet id.
2. Drop extra id column
3. Drop columns **retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp** 
4. Combine **Doggo, floofer, pupper, puppo** columns into one column **dog_stage**

##### Code

In [8]:
# Set column width to infinite so entire content of 'text' column is displayed
pd.set_option('display.max_colwidth', -1)

In [9]:
# Extract dog stage from text
df_tweets_copy['dog_stage'] = df_tweets_copy['text'].str.lower().str.extract('(doggo|floofer|pupper|puppo)', expand=True)

In [10]:
# Drop extra columns
df_tweets_copy = df_tweets_copy.drop(
    ['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 
     'doggo', 'floofer', 'pupper', 'puppo'], axis=1)

In [ ]:
df_tweets_copy.sample(10)

In [ ]:
df_tweets_copy['dog_stage'].value_counts()

In [11]:
# Merge three tables on same id
df_clean = pd.merge(df_tweets_copy, df_tweets_api_copy, left_on='tweet_id', right_on='id', how='inner')
df_clean = pd.merge(df_clean, df_image_copy, on='tweet_id', how='inner')

In [12]:
# Drop extra id column
df_clean = df_clean.drop(['id'], axis = 1)

##### Test

In [74]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2068 entries, 0 to 2067
Data columns (total 24 columns):
tweet_id                 2068 non-null int64
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                2068 non-null object
source                   2068 non-null object
text                     2068 non-null object
expanded_urls            2068 non-null object
rating_numerator         2068 non-null int64
rating_denominator       2068 non-null int64
name                     2068 non-null object
dog_stage                357 non-null object
retweet_count            2068 non-null int64
favorite_count           2068 non-null int64
jpg_url                  2068 non-null object
img_num                  2068 non-null int64
p1                       2068 non-null object
p1_conf                  2068 non-null float64
p1_dog                   2068 non-null bool
p2                       2068 non-null object
p2_conf                  2068

After merging the tables, we got 2068 records and it solves the original missing records issue.
1. `df_image` had missing records (2,075 instead of 2,356)
2. `df_tweets_api` had missing records (2,343 instead of 2,356)

#### Split timestamp to year, month, day of week

##### Define
1. Change timestamp datatype 
2. create new columns year, month, day of week

In [28]:
# Convert Object type to datetime type
df_clean['timestamp'] =  pd.to_datetime(df_clean['timestamp'], format='%Y%m%d %H:%M:%S')

In [14]:
# Get year, month and dayOfWeek from timestamp and store in new columns
df_clean['year'] = df_clean['timestamp'].dt.year
df_clean['month'] = df_clean['timestamp'].dt.month
df_clean['dow'] = df_clean['timestamp'].dt.dayofweek

##### Test

In [15]:
df_clean.groupby(["year", "month"]).size()

year  month
2015  11       298
      12       367
2016  1        169
      2        111
      3        121
      4        54 
      5        57 
      6        81 
      7        93 
      8        64 
      9        71 
      10       71 
      11       68 
      12       61 
2017  1        77 
      2        68 
      3        51 
      4        44 
      5        45 
      6        45 
      7        50 
      8        2  
dtype: int64

In [146]:
df_clean['dow'].value_counts()

0    326
2    297
1    295
3    284
4    282
6    256
5    251
Name: dow, dtype: int64

### Quality
#### Issue1: Extra Data
`df_tweets`: Extra retweet records

##### Define
1. Filter out the retweet data from `df_clean` DataFrame 

##### Code

In [36]:
# Exclude retweet data from DataFrame df_tweets_copy
df_clean = df_clean[~df_clean['text'].map(lambda txt: txt.lstrip().startswith("RT"))]

##### Test

In [17]:
# Confirm retweet data is gone
len(df_clean[df_clean['text'].map(lambda txt: txt.lstrip().startswith("RT"))])

0

In [76]:
len(df_clean)

1992

####  Issue2: Missing Data

`df_tweets`: **expanded_urls** (2,297 instead of 2,356) 

##### Define
Make sure `df_tweets_api_copy` DataFrame has all 1,992 tweets ID in `df_tweets_copy`


##### Code

In [20]:
sum(df_clean['expanded_urls'].isnull())

0

##### Test
There is no missing `expanded_urls` after combinning datasets.

####  Issue3: Wrong dog names

##### Define
Get correct dog names

##### Code

In [37]:
# Check dog name with non-capitalized texts
df_clean[df_clean['name'].str.contains('^[a-z]\w*')][['tweet_id', 'text', 'name']]

tweet_id  \
21    887517139158093824   
50    881536004380872706   
139   859196978902773760   
153   855459453768019968   
289   828650029636317184   
434   806219024703037440   
519   792913359805018113   
642   772581559778025472   
658   770655142660169732   
686   765395769549590528   
751   755206590534418437   
805   748977405889503236   
809   748692773788876800   
810   748575535303884801   
817   747885874273214464   
819   747816857231626240   
828   746872823977771008   
833   746369468511756288   
839   745422732645535745   
854   743222593470234624   
866   741067306818797568   
874   740214038584557568   
897   736225175608430592   
919   731156023742988288   
920   730924654643314689   
937   728035342121635841   
990   717537687239008257   
1000  715733265223708672   
1045  710272297844797440   
1117  704859558691414016   
...                  ...   
1876  669564461267722241   
1905  668955713004314625   
1912  668815180734689280   
1918  668636665813057536   
1925  668614819948453888   
1931  668507509523615744   
1935  668466899341221888   
1948  668171859951755264   
1962  667861340749471744   
1968  667773195014021121   
1977  667538891197542400   
1986  667470559035432960   
2000  667177989038297088   
2016  666983947667116034   
2023  666781792255496192   
2026  666701168228331520   
2038  666411507551481857   
2039  666407126856765440   
2045  666337882303524864   
2046  666293911632134144   
2047  666287406224695296   
2057  666063827256086533   
2058  666058600524156928   
2059  666057090499244032   
2060  666055525042405380   
2061  666051853826850816   
2062  666050758794694657   
2064  666044226329800704   
2065  666033412701032449   
2066  666029285002620928   

                                                                                                                                                                    text  \
21    I've yet to rate a Venezuelan Hover Wiener. This is such an honor. 14/10 paw-inspiring af (IG: roxy.thedoxy) https://t.co/20VrLAA8ba                                 
50    Here is a pupper approaching maximum borkdrive. Zooming at never before seen speeds. 14/10 paw-inspiring af \n(IG: puffie_the_chow) https://t.co/ghXBIIeQZF          
139   We only rate dogs. This is quite clearly a smol broken polar bear. We'd appreciate if you only send dogs. Thank you... 12/10 https://t.co/g2nSyGenG9                 
153   Guys, we only rate dogs. This is quite clearly a bulbasaur. Please only send dogs. Thank you... 12/10 human used pet, it's super effective https://t.co/Xc7uj1C64x   
289   Occasionally, we're sent fantastic stories. This is one of them. 14/10 for Grace https://t.co/bZ4axuH6OK                                                             
434   We only rate dogs. Please stop sending in non-canines like this Freudian Poof Lion. This is incredibly frustrating... 11/10 https://t.co/IZidSrBvhi                  
519   Here is a perfect example of someone who has their priorities in order. 13/10 for both owner and Forrest https://t.co/LRyMrU7Wfq                                     
642   Guys this is getting so out of hand. We only rate dogs. This is a Galapagos Speed Panda. Pls only send dogs... 10/10 https://t.co/8lpAGaZRFn                         
658   We only rate dogs. Pls stop sending in non-canines like this Arctic Floof Kangaroo. This is very frustrating. 11/10 https://t.co/qlUDuPoE3d                          
686   This is my dog. Her name is Zoey. She knows I've been rating other dogs. She's not happy. 13/10 no bias at all https://t.co/ep1NkYoiwB                               
751   This is one of the most inspirational stories I've ever come across. I have no words. 14/10 for both doggo and owner https://t.co/I5ld3eKD5k                         
805   What jokester sent in a pic without a dog in it? This is not @rock_rates. This is @dog_rates. Thank you ...10/10 https://t.co/nDPaYHrtNX                             
809   That is Quizno. This is his beach. He

From the observation, I found some texts have "named" or "name is" that could be useful to extract the correct names.

In [38]:
# Get dog name from texts that have "named"
ix_named = df_clean['text'].str.contains('named')
df_clean.loc[ix_named, 'name'] = df_clean[ix_named]['text'].str.extract('named\s([A-Z]\w+)', expand = False)

In [39]:
# Get dog name from texts that have "name is"
ix_name_is = df_clean['text'].str.contains('name is')
df_clean.loc[ix_name_is, 'name'] = df_clean[ix_name_is]['text'].str.extract('name\sis\s([A-Z]\w+)', expand = False)

In [40]:
# Set dog name to NaN if name is not start with capitalized character or name is None
ix_nan = df_clean['name'].str.contains('^[a-z]\w*|None')
df_clean.loc[ix_nan, 'name'] = np.NaN

##### Test

In [41]:
# Make sure name is NaN when no name mentioned in the text
df_clean[['tweet_id', 'text', 'name']]

tweet_id  \
0     892420643555336193   
1     892177421306343426   
2     891815181378084864   
3     891689557279858688   
4     891327558926688256   
5     891087950875897856   
6     890971913173991426   
7     890729181411237888   
8     890609185150312448   
9     890240255349198849   
10    890006608113172480   
11    889880896479866881   
12    889665388333682689   
13    889638837579907072   
14    889531135344209921   
15    889278841981685760   
16    888917238123831296   
17    888804989199671297   
18    888554962724278272   
19    888078434458587136   
20    887705289381826560   
21    887517139158093824   
22    887473957103951883   
23    887343217045368832   
24    887101392804085760   
25    886983233522544640   
26    886736880519319552   
27    886680336477933568   
28    886366144734445568   
29    886258384151887873   
...                  ...   
2038  666411507551481857   
2039  666407126856765440   
2040  666396247373291520   
2041  666373753744588802   
2042  666362758909284353   
2043  666353288456101888   
2044  666345417576210432   
2045  666337882303524864   
2046  666293911632134144   
2047  666287406224695296   
2048  666273097616637952   
2049  666268910803644416   
2050  666104133288665088   
2051  666102155909144576   
2052  666099513787052032   
2053  666094000022159362   
2054  666082916733198337   
2055  666073100786774016   
2056  666071193221509120   
2057  666063827256086533   
2058  666058600524156928   
2059  666057090499244032   
2060  666055525042405380   
2061  666051853826850816   
2062  666050758794694657   
2063  666049248165822465   
2064  666044226329800704   
2065  666033412701032449   
2066  666029285002620928   
2067  666020888022790149   

                                                                                                                                                                        text  \
0     This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                            
1     This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV       
2     This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                        
3     This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                                  
4     This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f       
5     Here we have a majestic great white breaching off South Africa's coast. Absolutely h*ckin breathtaking. 13/10 (IG: tucker_marlo) #BarkWeek https://t.co/kQ04fDDRmh       
6     Meet Jax. He enjoys ice cream so much he gets nervous around it. 13/10 help Jax enjoy more things by clicking below\n\nhttps://t.co/Zr4hWfAs1H https://t.co/tVJBRMnhxl   
7     When you watch your owner call another dog a good boy but then they turn back to you and say you're a great boy. 13/10 https://t.co/v0nONBcwxq                           
8     This is Zoey. She doesn't want to be one of the scary sharks. Just wants to be a snuggly pettable boatpet. 13/10 #BarkWeek https://t.co/9TwLuAGH0b                       
9     This is Cassie. She is a college pup. Studying international doggo communication and stick theory. 14/10 so elegant much sophisticate https://t.co/t1bfwz5S2A            
10    This is Koda. He is a South Australian deckshark. Deceptively deadly. Frighteningly majestic. 13/10 would risk a petting #BarkWeek https://t.co/dVPW0B0Mme               
11    This is Bruno. He is a service shark. Only gets out of the water to assist you. 13/10 terrifyingly good boy https://t.co/u1XPQMl29g                              

####  Issue 4: Some decimal ratings are not correct

##### Define
Correct the decimal rating

##### Code

In [42]:
# Use regex to get decimal rating
df_clean['new_rating'] = df_clean['text'].str.extract('(\d+\.?\d*\/\d*\.?\d+)', expand=True)

##### Test

In [43]:
df_clean['new_rating'].value_counts()

12/10       449
10/10       418
11/10       396
13/10       261
9/10        150
8/10        95 
7/10        51 
14/10       35 
6/10        32 
5/10        32 
3/10        19 
4/10        15 
2/10        9  
1/10        4  
0/10        2  
144/120     1  
1776/10     1  
50/50       1  
24/7        1  
11.26/10    1  
13.5/10     1  
60/50       1  
9/11        1  
121/110     1  
9.75/10     1  
84/70       1  
99/90       1  
204/170     1  
80/80       1  
1/2         1  
88/80       1  
165/150     1  
420/10      1  
44/40       1  
4/20        1  
143/130     1  
7/11        1  
11.27/10    1  
45/50       1  
Name: new_rating, dtype: int64

####  Issue 5: Multiple matches for fraction

##### Define
Get correct rating from multiple matches

##### Code

In [44]:
# Get dataframe that only have multiple matches
df_multi_match = df_clean[df_clean['text'].str.contains(r'(\d+\.?\d*\/\d*\.?\d+.+\d+\.?\d*\/\d*\.?\d+)')]

C:\Users\yeh\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  


In [45]:
df_multi_match = df_multi_match[['tweet_id', 'text', 'new_rating']]

In [46]:
# import the re library to get multiple matches for same index
import re
pat = re.compile("(?P<new_rating1>\d+\.?\d*\/\d*\.?\d+)")
df_multi_match['text'].str.extractall(pat).reset_index('match', drop=True)

new_rating1
611   12/10     
611   11/10     
820   10/10     
820   7/10      
871   9/11      
871   14/10     
962   4/20      
962   13/10     
996   50/50     
996   11/10     
1012  10/10     
1012  8/10      
1134  9/10      
1134  2/10      
1224  4/10      
1224  13/10     
1229  10/10     
1229  5/10      
1264  5/10      
1264  10/10     
1290  11/10     
1290  10/10     
1400  7/11      
1400  10/10     
1522  10/10     
1522  11/10     
1556  10/10     
1556  7/10      
1618  10/10     
1618  4/10      
1622  5/10      
1622  8/10      
1687  8/10      
1687  11/10     
1727  10/10     
1727  7/10      
1727  12/10     
1780  11/10     
1780  8/10      
1829  11/10     
1829  8/10      
1892  10/10     
1892  7/10      
1929  8/10      
1929  1/10      
1976  10/10     
1976  4/10      
1985  7/10      
1985  8/10      
2018  10/10     
2018  10/10     
2047  1/2       
2047  9/10

In [47]:
# Reshape the dataframe to have multiple rating columns for same record

df_multi_match = pd.concat([
    df_multi_match,
    (
        df_multi_match['text'].str.extractall(pat)
          .reset_index('match')          
    )
], axis=1).fillna('')

df_multi_match = df_multi_match[df_multi_match['match'] == 1].drop(['match'], axis = 1)
df_multi_match

tweet_id  \
611   777684233540206592   
820   747600769478692864   
871   740373189193256964   
962   722974582966214656   
996   716439118184652801   
1012  714258258790387713   
1134  703356393781329922   
1224  695064344191721472   
1229  694352839993344000   
1264  691483041324204033   
1290  689835978131935233   
1400  682962037429899265   
1522  677314812125323265   
1556  676191832485810177   
1618  674737130913071104   
1622  674646392044941312   
1687  673295268553605120   
1727  672248013293752320   
1780  671154572044468225   
1829  670434127938719744   
1892  669037058363662336   
1929  668537837512433665   
1976  667544320556335104   
1985  667491009379606528   
2018  666835007768551424   
2047  666287406224695296   

                                                                                                                                                     text  \
611   "Yep... just as I suspected. You're not flossing." 12/10 and 11/10 for the pup not flossing https://t.co/SuXcI9B7pQ                                   
820   This is Bookstore and Seaweed. Bookstore is tired and Seaweed is an asshole. 10/10 and 7/10 respectively https://t.co/eUGjGjjFVJ                      
871   After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ          
962   Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a                                                                                      
996   This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq                                 
1012  Meet Travis and Flurp. Travis is pretty chill but Flurp can't lie down properly. 10/10 &amp; 8/10\nget it together Flurp https://t.co/Akzl5ynMmE      
1134  This is Socks. That water pup w the super legs just splashed him. Socks did not appreciate that. 9/10 and 2/10 https://t.co/8rc5I22bBf                
1224  This may be the greatest video I've ever been sent. 4/10 for Charles the puppy, 13/10 overall. (Vid by @stevenxx_) https://t.co/uaJmNgXR2P            
1229  Meet Oliviér. He takes killer selfies. Has a dog of his own. It leaps at random &amp; can't bark for shit. 10/10 &amp; 5/10 https://t.co/6NgsQJuSBJ   
1264  When bae says they can't go out but you see them with someone else that same night. 5/10 &amp; 10/10 for heartbroken pup https://t.co/aenk0KpoWM      
1290  Meet Fynn &amp; Taco. Fynn is an all-powerful leaf lord and Taco is in the wrong place at the wrong time. 11/10 &amp; 10/10 https://t.co/MuqHPvtL8c   
1400  This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5           
1522  Meet Tassy &amp; Bee. Tassy is pretty chill, but Bee is convinced the Ruffles are haunted. 10/10 &amp; 11/10 respectively https://t.co/fgORpmTN9C     
1556  These two pups just met and have instantly bonded. Spectacular scene. Mesmerizing af. 10/10 and 7/10 for blue dog https://t.co/gwryaJO4tC             
1618  Meet Rufio. He is unaware of the pink legless pupper wrapped around him. Might want to get that checked 10/10 &amp; 4/10 https://t.co/KNfLnYPmYh      
1622  Two gorgeous dogs here. Little waddling dog is a rebel. Refuses to look at camera. Must be a preteen. 5/10 &amp; 8/10 https://t.co/YPfw7oahbD         
1687  Meet Eve. She's a raging alcoholic 8/10 (would b 11/10 but pupper alcoholism is a tragic issue that I can't condone) https://t.co/U36HYQIijg          
1727  10/10 for dog. 7/10 for cat. 12/10 for human. Much skill. Would pet all https://t.co/uhx5gfpx5k                                                       
1780  Meet Holly. She's trying to teach small human-like pup about blocks but he's not paying attention smh. 11/10 &amp; 8/10 https://t.co/RcksaUrGNu       
1829  Meet Hank and Sully. Hank is very proud of the pumpkin they found and Sully doesn't give a shit. 11/10 and 8/10 h

In [48]:
# function to get right rating into new_rating column
def getRightRating(ix_list, columnName):
    for i in ix_list:
        ix = df_multi_match.index == i
        df_multi_match.loc[ix, 'new_rating'] = df_multi_match.loc[ix, columnName]

In [49]:
# According to observation, some records have more than one dog rating and the first one will be used for now
# Some records have multiple matches but should only have one rating. 
# If new_rating1 value is correct, new_rating will be reokaced by new_rating1
getRightRating([871, 962, 996, 1400, 2047], 'new_rating1')

In [50]:
df_multi_match

tweet_id  \
611   777684233540206592   
820   747600769478692864   
871   740373189193256964   
962   722974582966214656   
996   716439118184652801   
1012  714258258790387713   
1134  703356393781329922   
1224  695064344191721472   
1229  694352839993344000   
1264  691483041324204033   
1290  689835978131935233   
1400  682962037429899265   
1522  677314812125323265   
1556  676191832485810177   
1618  674737130913071104   
1622  674646392044941312   
1687  673295268553605120   
1727  672248013293752320   
1780  671154572044468225   
1829  670434127938719744   
1892  669037058363662336   
1929  668537837512433665   
1976  667544320556335104   
1985  667491009379606528   
2018  666835007768551424   
2047  666287406224695296   

                                                                                                                                                     text  \
611   "Yep... just as I suspected. You're not flossing." 12/10 and 11/10 for the pup not flossing https://t.co/SuXcI9B7pQ                                   
820   This is Bookstore and Seaweed. Bookstore is tired and Seaweed is an asshole. 10/10 and 7/10 respectively https://t.co/eUGjGjjFVJ                      
871   After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ          
962   Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a                                                                                      
996   This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq                                 
1012  Meet Travis and Flurp. Travis is pretty chill but Flurp can't lie down properly. 10/10 &amp; 8/10\nget it together Flurp https://t.co/Akzl5ynMmE      
1134  This is Socks. That water pup w the super legs just splashed him. Socks did not appreciate that. 9/10 and 2/10 https://t.co/8rc5I22bBf                
1224  This may be the greatest video I've ever been sent. 4/10 for Charles the puppy, 13/10 overall. (Vid by @stevenxx_) https://t.co/uaJmNgXR2P            
1229  Meet Oliviér. He takes killer selfies. Has a dog of his own. It leaps at random &amp; can't bark for shit. 10/10 &amp; 5/10 https://t.co/6NgsQJuSBJ   
1264  When bae says they can't go out but you see them with someone else that same night. 5/10 &amp; 10/10 for heartbroken pup https://t.co/aenk0KpoWM      
1290  Meet Fynn &amp; Taco. Fynn is an all-powerful leaf lord and Taco is in the wrong place at the wrong time. 11/10 &amp; 10/10 https://t.co/MuqHPvtL8c   
1400  This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5           
1522  Meet Tassy &amp; Bee. Tassy is pretty chill, but Bee is convinced the Ruffles are haunted. 10/10 &amp; 11/10 respectively https://t.co/fgORpmTN9C     
1556  These two pups just met and have instantly bonded. Spectacular scene. Mesmerizing af. 10/10 and 7/10 for blue dog https://t.co/gwryaJO4tC             
1618  Meet Rufio. He is unaware of the pink legless pupper wrapped around him. Might want to get that checked 10/10 &amp; 4/10 https://t.co/KNfLnYPmYh      
1622  Two gorgeous dogs here. Little waddling dog is a rebel. Refuses to look at camera. Must be a preteen. 5/10 &amp; 8/10 https://t.co/YPfw7oahbD         
1687  Meet Eve. She's a raging alcoholic 8/10 (would b 11/10 but pupper alcoholism is a tragic issue that I can't condone) https://t.co/U36HYQIijg          
1727  10/10 for dog. 7/10 for cat. 12/10 for human. Much skill. Would pet all https://t.co/uhx5gfpx5k                                                       
1780  Meet Holly. She's trying to teach small human-like pup about blocks but he's not paying attention smh. 11/10 &amp; 8/10 https://t.co/RcksaUrGNu       
1829  Meet Hank and Sully. Hank is very proud of the pumpkin they found and Sully doesn't give a shit. 11/10 and 8/10 h

In [51]:
# replace the df_clean DataFrame by the df_multi_match DataFrame for correct rating
ix_multi = df_clean['text'].str.contains(r'(\d+\.?\d*\/\d*\.?\d+.+\d+\.?\d*\/\d*\.?\d+)')
df_clean.loc[ix_multi, 'new_rating'] = df_multi_match[ix_multi]['new_rating']

C:\Users\yeh\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  
C:\Users\yeh\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  This is separate from the ipykernel package so we can avoid doing imports until


In [52]:
# Make sure df_clean DataFrame is replaced by correct ratings
df_clean.loc[ix_multi, 'new_rating']

611     12/10
820     10/10
871     14/10
962     13/10
996     11/10
1012    10/10
1134    9/10 
1224    4/10 
1229    10/10
1264    5/10 
1290    11/10
1400    10/10
1522    10/10
1556    10/10
1618    10/10
1622    5/10 
1687    8/10 
1727    10/10
1780    11/10
1829    11/10
1892    10/10
1929    8/10 
1976    10/10
1985    7/10 
2018    10/10
2047    9/10 
Name: new_rating, dtype: object

In [53]:
# Split new_rating to numerator and denomator
df_clean['rating_numerator'] = list(map(lambda x: x.split('/')[0], df_clean['new_rating']))
df_clean['rating_denominator'] = list(map(lambda x: x.split('/')[1], df_clean['new_rating']))
df_clean = df_clean.drop('new_rating', axis =1)

##### Test

In [54]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1992 entries, 0 to 2067
Data columns (total 27 columns):
tweet_id                 1992 non-null int64
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                1992 non-null datetime64[ns]
source                   1992 non-null object
text                     1992 non-null object
expanded_urls            1992 non-null object
rating_numerator         1992 non-null object
rating_denominator       1992 non-null object
name                     1378 non-null object
dog_stage                342 non-null object
retweet_count            1992 non-null int64
favorite_count           1992 non-null int64
jpg_url                  1992 non-null object
img_num                  1992 non-null int64
p1                       1992 non-null object
p1_conf                  1992 non-null float64
p1_dog                   1992 non-null bool
p2                       1992 non-null object
p2_conf            

In [32]:
df_clean['rating_numerator'].value_counts()

12       449
10       419
11       397
13       262
9        151
8        95 
7        51 
14       36 
5        32 
6        32 
3        19 
4        15 
2        9  
1        4  
0        2  
44       1  
144      1  
121      1  
420      1  
24       1  
1776     1  
80       1  
13.5     1  
84       1  
60       1  
9.75     1  
165      1  
204      1  
45       1  
11.26    1  
11.27    1  
99       1  
143      1  
88       1  
Name: rating_numerator, dtype: int64

In [33]:
df_clean['rating_denominator'].value_counts()

10     1979
50     2   
80     2   
70     1   
150    1   
40     1   
90     1   
120    1   
7      1   
130    1   
170    1   
110    1   
Name: rating_denominator, dtype: int64

####  Issue 6: Rating denominator not equal to 10

##### Define
Make sure the rating is correct even though the denominator is not 10

##### Code

In [55]:
df_clean[df_clean['rating_denominator'] != '10'][['tweet_id', 'text', 'rating_denominator']]

tweet_id  \
341   820690176645140481   
411   810984652412424192   
730   758467244762497024   
919   731156023742988288   
1017  713900603437621249   
1042  710658690886586372   
1060  709198395643068416   
1126  704054845121142784   
1202  697463031882764288   
1374  684225744407494656   
1375  684222868335505415   
1507  677716515794329600   
1566  675853064436391936   

                                                                                                                                                                           text  \
341   The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd                                                                         
411   Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx   
730   Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE                                                                                              
919   Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv                                                   
1017  Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1                                                                               
1042  Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12                                                                
1060  From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK                          
1126  Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa                                                                                          
1202  Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ                                                                                
1374  Two sneaky puppers were not initially seen, moving the rating to 143/130. Please forgive us. Thank you https://t.co/kRK51Y5ac3                                              
1375  Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110 https://t.co/1zfnTJLt55                                
1507  IT'S PUPPERGEDDON. Total of 144/120 ...I think https://t.co/ZanVtAtvIq                                                                                                      
1566  Here we have an entire platoon of puppers. Total score: 88/80 would pet all at once https://t.co/y93p6FLvVw                                                                 

     rating_denominator  
341   70                 
411   7                  
730   150                
919   170                
1017  90                 
1042  80                 
1060  50                 
1126  50                 
1202  40                 
1374  130                
1375  110                
1507  120                
1566  80

In [56]:
# There is no rating in index 411 record, so drop it 
df_clean = df_clean.drop(df_clean[df_clean['tweet_id'] == 810984652412424192].index)
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1991 entries, 0 to 2067
Data columns (total 27 columns):
tweet_id                 1991 non-null int64
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                1991 non-null datetime64[ns]
source                   1991 non-null object
text                     1991 non-null object
expanded_urls            1991 non-null object
rating_numerator         1991 non-null object
rating_denominator       1991 non-null object
name                     1377 non-null object
dog_stage                342 non-null object
retweet_count            1991 non-null int64
favorite_count           1991 non-null int64
jpg_url                  1991 non-null object
img_num                  1991 non-null int64
p1                       1991 non-null object
p1_conf                  1991 non-null float64
p1_dog                   1991 non-null bool
p2                       1991 non-null object
p2_conf            

####  Issue 7: P1, P2, P3 should have consistent capitalization

##### Define
Transform non-capitalized values

##### Code

In [57]:
df_clean['p1'] = df_clean['p1'].str.capitalize()
df_clean['p2'] = df_clean['p2'].str.capitalize()
df_clean['p3'] = df_clean['p3'].str.capitalize()

##### Test

In [58]:
df_clean[['p1', 'p2', 'p3']]

p1                        p2  \
0     Orange                       Bagel                      
1     Chihuahua                    Pekinese                   
2     Chihuahua                    Malamute                   
3     Paper_towel                  Labrador_retriever         
4     Basset                       English_springer           
5     Chesapeake_bay_retriever     Irish_terrier              
6     Appenzeller                  Border_collie              
7     Pomeranian                   Eskimo_dog                 
8     Irish_terrier                Irish_setter               
9     Pembroke                     Cardigan                   
10    Samoyed                      Pomeranian                 
11    French_bulldog               Labrador_retriever         
12    Pembroke                     Cardigan                   
13    French_bulldog               Boxer                      
14    Golden_retriever             Labrador_retriever         
15    Whippet                      Borzoi                     
16    Golden_retriever             Tibetan_mastiff            
17    Golden_retriever             Labrador_retriever         
18    Siberian_husky               Eskimo_dog                 
19    French_bulldog               Pug                        
20    Basset                       Redbone                    
21    Limousine                    Tow_truck                  
22    Pembroke                     Rhodesian_ridgeback        
23    Mexican_hairless             Sea_lion                   
24    Samoyed                      Eskimo_dog                 
25    Chihuahua                    Toy_terrier                
26    Kuvasz                       Great_pyrenees             
27    Convertible                  Sports_car                 
28    French_bulldog               Chihuahua                  
29    Pug                          Shower_cap                 
...   ...                                 ...                 
2038  Coho                         Barracouta                 
2039  Black-and-tan_coonhound      Bloodhound                 
2040  Chihuahua                    Toy_terrier                
2041  Soft-coated_wheaten_terrier  Afghan_hound               
2042  Guinea_pig                   Skunk                      
2043  Malamute                     Siberian_husky             
2044  Golden_retriever             Chesapeake_bay_retriever   
2045  Ox                           Newfoundland               
2046  Three-toed_sloth             Otter                      
2047  Maltese_dog                  Toy_poodle                 
2048  Italian_greyhound            Toy_terrier                
2049  Desktop_computer             Desk                       
2050  Hen                          Cock                       
2051  English_setter               Newfoundland               
2052  Lhasa                        Shih-tzu                   
2053  Bloodhound                   German_shepherd            
2054  Pug                          Bull_mastiff               
2055  Walker_hound                 English_foxhound           
2056  Gordon_setter                Yorkshire_terrier          
2057  Golden_retriever             Tibetan_mastiff            
2058  Miniature_poodle             Komondor                   
2059  Shopping_cart                Shopping_basket            
2060  Chow                         Tibetan_mastiff            
2061  Box_turtle                   Mud_turtle                 
2062  Bernese_mountain_dog         English_springer           
2063  Miniature_pinscher           Rottweiler                 
2064  Rhodesian_ridgeback          Redbone                    
2065  German_shepherd              Malinois                   
2066  Redbone                      Miniature_pinscher         
2067  Welsh_springer_spaniel       Collie                     

                               p3  
0     Banana                       
1     Papillon                     
2     Kelpie    

In [59]:
# Create the twitter_archive_master.csv
df_clean.to_csv("twitter_archive_master.csv")